rust/main/chains/hyperlane-cosmos/src/providers/cosmos/provider.rs

```diff
@@ -197,8 +202,23 @@ impl CosmosProvider {
     }
 
     /// Extract contract address from transaction.
-    /// Assumes that there is only one `MsgExecuteContract` message in the transaction
     fn contract(tx: &Tx, tx_hash: &H256) -> ChainResult<H256> {
+        // We merge two error messages together so that both of them are reported
+        match Self::contract_address_from_msg_execute_contract(tx, tx_hash) {
+            Ok(contract) => Ok(contract),
+            Err(mece) => match Self::contract_address_from_msg_recv_packet(tx, tx_hash) {
+                Ok(contract) => Ok(contract),
+                Err(mrpe) => {
+                    let msg = mece.to_string() + " and " + &mrpe.to_string();
+                    warn!(?tx_hash, msg);
+                    Err(ChainCommunicationError::CustomError(msg.to_owned()))?
```

Comment:
I don't feel too good about manually stitching together error strings, and I think this is a good chance to add a quick fix the other rushed error types in this file. How about we:
- define a new `HyperlaneCosmosError::ParsingFailed(String)` field and have the parser functions like `contract_address_from_msg_execute_contract` return a `Result<..., HyperlaneCosmosError>` instead of the `Custom` error
- also define a new `HyperlaneCosmosError::ParsingAttemptsFailed(Vec<HyperlaneCosmosError>)`, such that the errors from the two parsing attempts above are aggregated in this vec, and how the entire error is ultimately formatted is up to the `Debug` implementation for `Vec`, which is in line with how we report fallback provider errors `HyperlaneCosmosError::FallbackProvidersFailed`

Reply:
It is great! I've implemented it.